### ISSUE TITLE
Improve design & refactor the Study search auto-complete

### ISSUE THREAD

**Issue by fabd** [fabd/kanji-koohii#9]
The Study search auto-complete UX needs to be improved on mobile and desktop. It's tiny and ugly.

AND it needs to be refactored completely with a VueJS 2 component, which makes much more sense, and allows to expand its functonality. The current code is extremely old Javascript that was made to work on IE6...

- Find a good Vue JS **2.x** auto-complete component with **no jQuery** dependencies. (Preferably a simple one that I can customize better than a full featured one which 90% of the options I won't need.)

- Hack the component so that it takes the data from the RTK 5th / 6th editions Javascript added in the page  by the backend based on the user's choice of RTK edition (we can't bake the auto complete data into the component itself, unless we baked in BOTH RTK edition indexes, which is not a big cost in terms of delivery size... but seems like the wrong solution for expandability)

- Optimize row size and font-size for mobile & touch

Labelled `nice-to-have` because it currently works.

**Comment by faneca** [comment]
Please don't forget that when you have lots of customized keywords (e.g. when you have them translated to another language or decided to resort to Japanese-only keywords), searching becomes confusing: the matching performed to show items in the dropdown is only executed against the original, by-default keywords; and if you happen to select one of them (which is relatively trivial to do inadvertently if you have a custom keyword sharing prefix with one of the original keywords), then you are taken to a search result where the search term was that of the original keyword, instead of what you typed (even if you chose on purpose one of the results shown on the dropdown, it's easy to end up in a "sorry, no results" page).

**Comment by fabd** [comment]
Hmm yeah short of actually completing custom keywords there is no ideal solution here since autocomplete will try to do its job.

Until now, I think you can type a space first to prevent the auto completion while entering a custom keyword?

Then there is the backend side where the search term needs to be correctly matches. I believe I made an update a while ago to check for an exact match first. Let me know if it's still broken.

**Thoughts**
- I prefer to avoid an ajax dropdown that hits the server for custom keywords
- a more complex solution which I've put off indefinitely is to recreate the javascript file with the user's own set of keywords, as a *cachable* resource, and update that whenever user edits keywords, but then besides caching & invalidating woes, the original keywords are not available in the dropdown

**Comment by faneca** [comment]
> I think you can type a space first to prevent the auto completion while entering a custom keyword?

So simple! Never thought of that :-$. Definitely gonna try it next time I log to the site, thank you!

Anyway, I have some
### Proposed changes
(which would make my life easier, don't seem too complicated to implement and -hopefully- won't break current behavior most of the time)

TL;DR
1. Don't select a dropdown option unless explicitly told to do so, and
2. Search by frame number instead of by keyword when the option *is* explicitly selected.

[rant]

> 
> I should probably start by saying I really never liked combo boxes and similar dropdowns. Not *per se*, they're a nice and powerful UI concept if done right, but overall the different implementations in the wild are horribly inconsistent with each other. Just the sitting-next-to-each-other location and search bars in (a lot of) Firefox (versions) should be enough of a case.
> 
> Some examples of the usual issues I have with them:
>   - What should happen (to the already typed textbox's content, to the *highlighted* value, to the *effectively selected* value, ...) when you press *tab*, *esc* or *enter*? (and what if the widget allows multiple selection? I'm glad this is not the case at hand)
>   - Should having an option already highlighted (or selected, which is not quite the same) change the expected behavior when you press any of those keys?
>   - When is it the best moment to display the dropdown? As soon as you focus the textbox? When you start typing? When you have typed at least three characters? Only when you explicitly ask it to by pressing *alt-down* or clicking on it? Is the *alt-down* hotkey really that ubiquitous? Does clicking work consistently?
> 
> 

[/rant]

Having taken that out of my system :-), a relatively easy-to-implement way to avoid most of the "confusing behavior" (at least IMHO) would be to leave the current functionality pretty much untouched, but with a subtle change: avoid the automatic selection of any option as you type (currently, the first one is "magically" chosen), and only do select one when the user either a) clicks on any option or b) presses tab or the vertical cursor arrows (to highlight one option) and then enter (to select it).

That way,
  - When you just type something and then press enter, even if you are shown some "suggestions" in the dropdown, you'll look up exactly what you wrote, not *the whole keyword* of the first match -- which may very well be absent from your custom keywords list, so you'll end up in an empty page with a "not found" message (though your above tip to make this work with the current implementation could solve this use case nicely... on a side note, please note that current behavior has a race condition which makes it give different results depending on the shown or hidden status of the dropdown).
  - Alternatively, if you press tab or the arrow keys (possibly multiple times) before pressing enter, you'll go directly to the study page you explicitly selected. For this to work correctly, too, when your keywords are heavily customized, it'd probably be best to do the lookup by frame number instead of by keyword, avoiding again the "not found" result.
  - If you forget to press tab or the arrows and just press enter after typing, triggering the behavior described in the previous paragraph (and, say, you didn't customize too much your keywords), given the current implementation of the search feature, chances are you'll end up in that same study page anyway.

### Final notes
I agree with the rest of your post: I gave it some thought in the past (maybe even wrote about it in the forum?) and never came up with a satisfying solution for the custom keywords search; you summarized pretty much everything I could imagine (just better ;-).

An alternative to having *your server* provide, in any way, the custom keywords search could be a contributed (and let's say "offline") tool which, given the (exported) custom keywords as input, could construct a user script to be injected in the study pages by GreaseMonkey and family. Admittedly, it would be somewhat unfriendly to setup (but better than nothing, at least for technically knowledgeable users). Nonetheless **I suspect the audience for this would be minimal**, hence no worth the effort (but if anyone is interested, I remember there was already a user script published somewhere which at some point provided this feature and which could be used as a base; unfortunately I've been spending way more than my currently available time for "pet projects" lately).

**Comment by fabd** [comment]
Regarding "race condition" or other bugs , the current auto complete is antiquated javascript. Plan is to use a VueJs component , perhaps [this Element UI Select](http://element.eleme.io/#/en-US/component/select#option-filtering) so the behaviour will be different. (as an aside it takes time actually to try them, and convert them since they often don't match my build or use tools I don't want, or include too many cruft that adds up to the Koohii Javascript bundle... also include bunch of dependencies I already have in other forms which is the challenge with an old codebase)

**Solutions**

I agree about the main suggestion to explicitly pick a suggestion from the drop down.

Then another improvement I could potentially do... is to return a page with a list of matches if count results > 1. It's one extra page to get to the kanji.. but later the logic could be refactored to an ajax UI of some sort.

**Comment by fabd** [comment]
@faneca Made a basic test of the vue-multiselect component, loaded with Koohii keywords. Have a look. This auto-complete matches *any* party of the keyword. Eg. "self" matches "selfish" and "myself" 

[Basic vue-multiselect with RTK keywords](http://codepen.io/fabd/pen/qRREaK)

Man, CodePen is cool. Great for discussing changes! I am investigating the options. Not too fond of the 22kb size, but the autocomplete is a big part of the Study page so it may be worth it.

**Comment by faneca** [comment]
> Meh.. another component not up to date for Vue 2. Abandoned.

I know exactly the feeling :-(. Also, 22k of javascript (even if gzipped is less than 7k) seems like a lot for just this feature. Nevermind, you have a lot of other issues to deal with and this is just a marginal case I can live with. I'd be extremely happy if the implementation of the two proposed enhancements kicks in (don't select from the dropdown implicitly and do a two-step search if there is no perfect match), but even without them I'd be fine. In any case, I appreciate your efforts a lot ;-)

[On a side note, an alternative (that I'm sure you're totally aware of) to the "cachable resource with a javascript list of custom keywords" could be to put that list in the localStorage, but in the end it's more or less the same problem, or even worse: the server still has to have all the custom keywords stored for the times when a user logs in with a new (or reinstalled) browser / device / profile (and send the entire list to the client in that cases), so the issues with invalidation and keeping track of the versions stored in each client persist and are far from trivial.]

**Comment by fabd** [comment]
Yep localStorage wouldn't help a lot here.

I'll probably end up making one with Vue. I want to get really skilled with Vue anyway. 

I think custom  is the way to go. [Here is a simple tutorial to make one.](http://taha-sh.com/blog/building-an-awesome-reusable-autocomplete-input-component-in-vue-21-part-one) . Maybe I'll make it on CodePen too that way people can help me out try different ideas.
complicated to build my own.

If I just replace the old one with a simple VueJs, at the very least it let people play with it in CodePen and maybe if something interesting comes out, I can merge it in.

The main improvement refactoring the old code is to be able to style the dropdown. Not just the main dimensions and color, but also to layout each item in the dropdown just like I want.

But right now as I announced on Gitter, I am going to push to live very soon because I'm losing focus and been postponing this for too long. I want to fix 2-3 things and then I'll push **staging** to live. So I'll play with the dropdown later.

**Comment by rinre** [comment]
Just wanted to say I agree with faneca's ideas on this one. I don't really understand the coding stuff like you guys, but it seems like making the autocomplete include custom keywords is a coding/server hell, right?
So with that off the table, faneca's suggested small behavioural changes would be wonderful (he described it better than I can, main thing would be not ending up in another kanji than you want to look up or in “not found” when the first option gets automatically selected).
Btw, the codepen example looks cool, shame it wasn't in the new version. 
Good job on the site as alqays, it's really going forward, so many issues closed already! ;-)

**Comment by fabd** [comment]
Tagged on hold and created a new issue for better task breakdown.

#114 **Refactor the study search dropdown to VueJS** (and ditch the "autocomplete" Javascript include).
